### PR TITLE
dusk-merkle: Validate archived opening positions

### LINKED ISSUE #121: Validate archived opening positions
## Summary

Derived `CheckBytes` validation only checks the structure of an archived `Opening`. It does not reject entries in `Opening::positions` that are outside the tree arity.

Although verification now rejects these positions, other consumers such as serialization and direct position access can still observe a semantically invalid opening. Validate the position invariant at the archive boundary without changing the public API or archive layout.

## Acceptance criteria

- Checked RKYV deserialization rejects every archived position greater than or equal to the tree arity.
- Malformed input returns an error without unwinding.
- Valid openings round-trip for supported archive sizes and continue to verify unchanged.
- Validation does not rely on consumer-specific guards.
- The public API and archive layout remain unchanged.

### PR DESCRIPTION
**Pull request by HDauven**
## Summary

- validate every archived opening position against the generic tree arity during checked RKYV deserialization
- return a validation error for malformed positions without unwinding
- add boundary-corruption and valid roundtrip regressions for 16-, 32-, and 64-bit RKYV archives
- document the fix in the changelog

## Validation

- `make test`
- `make no-std`
- `make cq`
- `make check`
- `make doc`
- `cargo test --release -p dusk-merkle --features=rkyv-impl,size_16`
- `cargo test --release -p dusk-merkle --features=rkyv-impl,size_32`
- `cargo test --release -p dusk-merkle --features=rkyv-impl,size_64`
- `cargo test --release -p poseidon-merkle --features=zk,rkyv-impl,size_16`
- `cargo test --release -p poseidon-merkle --features=zk,rkyv-impl,size_64`

## Compatibility

The public API and archive layout are unchanged. Valid openings roundtrip and verify as before; checked deserialization now rejects semantically invalid positions at the archive boundary.

Resolves #121

### CODE REVIEW

File: dusk-merkle/src/opening.rs [hunk 1]
```diff
@@ -16,17 +14,111 @@ use crate::{Aggregate, Node, Tree, init_array};
...
+    impl<C, T, const H: usize, const A: usize> CheckBytes<C>
+        for ArchivedOpening<T, H, A>
+    where
+        C: ?Sized,
+        T: Archive,
+        Archived<T>: CheckBytes<C>,
+        Archived<[[T; A]; H]>: CheckBytes<C>,
+        Archived<[usize; H]>: CheckBytes<C>,
+    {
+        type Error = StructCheckError;
```
Comment (moCello) [inline dusk-merkle/src/opening.rs:61]: **NOTE:** A field later added to `Opening` compiles without being checked here, silently reopening the gap this PR closes.

**Context:** The derive regenerated per-field checks on every build; the manual impl enumerates `root`, `branch`, `positions` by hand, and nothing fails when that list goes stale. `check_archived_root` is the soundness gate for zero-copy access, so for a `T` with archived-type invariants an unchecked field is undefined behavior, not just a missed semantic check. Today's three fields are all plain integers or delegate to their own impls, so nothing is wrong now — what the change gives up is the derive's automatic tracking, and this buys it back.

**Fix:** Make the enumeration compiler-enforced: end `check_bytes` with an exhaustive pattern on the checked reference and no `..`, so adding a field breaks the build until this impl is updated. Two lines before the final `Ok`, destructuring the reference rather than the raw pointer:

```rust
let checked = unsafe { &*value };
let Self { root: _, branch: _, positions: _ } = checked;
Ok(checked)
```

File: dusk-merkle/src/opening.rs [hunk 2]
```diff
@@ -302,4 +394,112 @@ mod tests {
...
+            // Include a value whose lower half is zero so narrowing to a
+            // smaller host `usize` cannot accidentally make it valid.
+            let high_bits_only =
+                (1u128 << (rkyv::FixedUsize::BITS / 2)) as rkyv::FixedUsize;
+            let invalid_positions = [
+                A as rkyv::FixedUsize,
+                (A + 1) as rkyv::FixedUsize,
+                high_bits_only,
+                rkyv::FixedUsize::MAX,
+            ];
```
Comment (moCello) [inline dusk-merkle/src/opening.rs:490]: **NOTE:** The width-dependent cases this test targets run only under manual `size_16`/`size_64` invocations — CI's `make test` pins `size_32`, so the fix lands there in a follow-up, not this PR.

**Context:** The narrowing this case guards against (high bits lost in a `usize` cast) is only observable where the target's `usize` is narrower than the archive width, e.g. `size_64` archives read on a 32-bit target such as wasm32. On a 64-bit runner even a `size_64` run still passes if the `u128` widening in `check_bytes` regresses to an `as usize` cast, so a feature matrix alone doesn't make the narrowing falsifiable.

**Fix:** File a tracking issue for the Makefile/CI: extend the test target to `--features=rkyv-impl,size_16` and `--features=rkyv-impl,size_64`, and run the `size_64` tests once on a 32-bit target to make the narrowing case falsifiable.

**Review by moCello — APPROVED** [review]
APPROVE — the manual `CheckBytes` impl is structurally faithful to what the derive emitted: same field order, same targets, same error type and field names, equivalent bounds, and strictly stronger, since the derive accepted `position == A` and this does not. The semantic check closes #121 exactly as the issue specifies, and the corruption tests cover the boundary and width edge cases. Nothing gates the merge.

Two notes below. The first is worth taking here rather than later: replacing the derive means giving up its automatic field tracking, and two lines of exhaustive destructuring at the end of `check_bytes` buy that back — cheap now, and the kind of thing that is only ever noticed once a field has already been added without a matching check. The second is genuinely a follow-up: making the narrowing case falsifiable needs a CI matrix change and a 32-bit target, neither of which belongs on this branch.

File: dusk-merkle/src/opening.rs (hunk 1 above)
Comment (HDauven) [reply dusk-merkle/src/opening.rs:61]: Done. The checked reference is now exhaustively destructured before return, so adding a field breaks the manual validator until it is updated.

File: dusk-merkle/src/opening.rs (hunk 2 above)
Comment (HDauven) [reply dusk-merkle/src/opening.rs:490]: Tracked separately in #124.
